### PR TITLE
Support HostAlias for HostNetwork Pods

### LINKED ISSUE #48398: HostAlias to Support HostNetwork Pods
Currently, HostAlias feature [does not support HostNetwork pods](https://github.com/kubernetes/kubernetes/blob/faf4e57f1b9c2ea4671f62c96d4470e4757a3a4e/pkg/api/validation/validation.go#L2062-L2070). This is because HostNetwork pods currently [do not](https://github.com/kubernetes/kubernetes/blob/48b3fb84abdc0fee12299eed262da287d48110de/pkg/kubelet/kubelet_pods.go#L112-L118) have its hosts files managed by kubelet.

Once https://github.com/kubernetes/kubernetes/issues/48397 resolves, we can change the validation to make HostAlias also supported by HostNetwork pods.

/kind feature
/sig node

@yujuhong @hongchaodeng @thockin

### PR DESCRIPTION
**Pull request by evie404**
**What this PR does / why we need it**: Currently, HostAlias does not support HostNetwork pods because historically, kubelet only manages hosts file for non-HostNetwork pods. With the recent change in https://github.com/kubernetes/kubernetes/pull/49140, kubelet now manages hosts file for all Pods, which enables HostAlias support also.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #48398

**Special notes for your reviewer**: might be easier to review commit-by-commit

**Release note**:

```release-note
HostAlias is now supported for both non-HostNetwork Pods and HostNetwork Pods.
```

@yujuhong @hongchaodeng @thockin

### CODE REVIEW

File: pkg/kubelet/kubelet_pods_test.go [hunk 1]
```diff
@@ -201,6 +213,70 @@ fe00::2	ip6-allrouters
 		},
 		{
 			"hosts_test_file2",
+			[]v1.HostAlias{},
+			`# another hosts file for testing.
```
Comment (hongchaodeng) [inline pkg/kubelet/kubelet_pods_test.go]: No need to this test? It's already tested in above?

File: pkg/kubelet/kubelet_pods_test.go [hunk 2]
```diff
@@ -184,11 +184,23 @@ func TestMakeMounts(t *testing.T) {
 
 func TestNodeHostsFileContent(t *testing.T) {
```
Comment (hongchaodeng) [inline pkg/kubelet/kubelet_pods_test.go]: This test seems to be only testing `nodeHostsFileContent()` ?
Any test to verify that `hostNetwork=true` hostAlias ?

**Review by hongchaodeng — COMMENTED** [review]
@rickypai Great work! Have some comments please take a look.

One question: I saw that `ensureHostsFile()` will overwrite the host `/etc/hosts` file. Is it an assumption that after kubelet starts no one will modify /etc/hosts ?

File: pkg/kubelet/kubelet_pods_test.go (hunk 2 above)
Comment (rickypai) [inline pkg/kubelet/kubelet_pods_test.go]: in `ensureHostsFile`, `nodeHostsFileContent` is only called when `hostNetwork=true`. so this is already testing the case.

File: pkg/kubelet/kubelet_pods_test.go (hunk 1 above)
Comment (rickypai) [inline pkg/kubelet/kubelet_pods_test.go]: tests different file path just to be exhaustive

**Comment by rickypai** [comment]
In the case `/etc/hosts` is modified after a hostNetwork pod is scheduled, the pod will not be able to get the new content. This is not ideal, but it is consistent with the old behavior when docker was responsible for `/etc/hosts`.

**Comment by hongchaodeng** [comment]
Yeah. I understand that it's the same as non-hostNetwork case.

Just to make sure that it's assumed that `/etc/hosts` is managed by docker and kubelet only, and further assumed that docker won't modify it after kubelet takes control. Isn't it?

**Comment by rickypai** [comment]
the original intention for kubelet-managed hosts file is to prevent docker from changing it. this was already the case for non-hostNetwork pods, so I think the assumption still applies for hostNetwork pods.

more background: https://kubernetes.io/docs/concepts/services-networking/add-entries-to-pod-etc-hosts-with-host-aliases/#why-does-kubelet-manage-the-hosts-file

**Comment by hongchaodeng** [comment]
👍

**Pull request merged**